RFC/src/RFC-0171_PeerToPeerMessagingProtocol.md

```diff
@@ -0,0 +1,662 @@
+PEER -->|has many| NA
+</div>
+
+#### NetAddress
+
+Represents one of the following:
+ 
+ - an IP address,
+ - an Onion address,
+ - or, an I2P address.
+ 
+```rust,compile_fail
+#[derive(Clone, PartialEq, Eq, Debug)]
+/// Represents an address which can be used to reach a node on the network
+pub enum NetAddress {
+    /// IPv4 and IPv6
+    IP(SocketAddress),
+    Tor(OnionAddress),
+    I2P(I2PAddress),
+}
+``` 
+
+#### MessageEnvelope
+
+Every Tari message MUST be wrapped in a MessageEnvelope. This represents the four frames of a multipart message
+which have just come off, or are about to go on to the wire. 
+
+Fields may include:
+
+| Name     | Frame | Length (octets) | Type      | Description                                                                                                                                                     |
+| -------- | ----- | --------------- | --------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| identity | 0     | 8               | `[u8;8]`  | The identifier that a `ZMQ_ROUTER` socket expects so that it knows the intended destination of the message. This can be thought of as a session token.          |
+| version  | 1     | 1               | `u8`      | The wire protocol version.                                                                                                                                      |
+| header   | 2     | as required     | `Vec<u8>` | Serialized bytes of data containing an unencrypted HeaderMessage.                                                                                               |
+| body     | 3     | as required     | `Vec<u8>` | Serialized bytes of data containing a unencrypted or encrypted message.                                                                                         |
+
+The header and decrypted body MUST be deserializable as per [RFC-0171: Message Serialisation](RFC-0171_MessageSerialisation.md)
+
+#### MessageHeader
+
+Every Tari message must have a header containing the following fields:
+
+| Name         | Type              | Description                                                                                        |
+| ------------ | ----------------- | -------------------------------------------------------------------------------------------------- |
+| version      | `String`          | The message protocol version (e.g. "Tari/1.0")."                                                   |
+| message_type | `MessageType`     | An enumeration of the message type of the body. As per [MessageType].                              |
+| source       | `PublicKey`       | The source public key.                                                                             |
+| dest         | `Option<NodeId>` | The destination [node ID]. This is optional.                                                       |
```

Comment:
This could be changed to have one of 3 types: None, NodeId or Public Key

Reply:
Do destination should be 

```rust
enum NodeDestination {
NodeId(NodeId),
PublicKey(PublicKey),
}
```

Is there a reason for having either rather than just one?